### ISSUE TITLE
Preview not working with Nuxt with trailing slashing because of Netlify

### ISSUE THREAD

**Issue by marvinhuebner** [storyblok/storyblok#613]
**The issue occurs while working with:** (check one with "x")   
- [x] *app.storyblok.com (CMS - Interface)*   

**Current behavior:**
Pages which are directly served by Netlify are served with a trailing slash. Read more about it here: https://dev.to/mornir/nuxt-netlify-and-the-trailing-slash-3gge

That's why we set the `trailingSlash` option to `true` to be consistent for SEO. This way the Storyblok preview can't find the page. 

**Expected behavior:**
I need a solution to show the preview in Storyblok.

**Steps to reproduce:**
Adding the `router.trailingSlash` in your `nuxt.config.js` and set it to `true`.
```js
router: {
    trailingSlash: true,
},
```

**Other information:**

- [Netlify + Nuxt - Remove trailing slash](https://answers.netlify.com/t/netlify-nuxt-remove-trailing-slash/14626)
- [Nuxt, Netlify and the trailing slash](https://dev.to/mornir/nuxt-netlify-and-the-trailing-slash-3gge)
- https://github.com/nuxt/nuxt.js/pull/7776#issuecomment-664215752

Thanks for your help in advance!

**Comment by marvinhuebner** [comment]
Maybe I have an idea for a workaround to solve this.

**My consideration:**
As a base I deploy a second version of the website on Netlify which I'm using for the editing preview in Storyblok.

I add an env variable `TRAILING_SLASH`, which I use to control the router option and I add a `urlHelper()` filter / function that generates the links with or without trailing slash.

```diff
router: {
-    trailingSlash: true,
+    trailingSlash: (process.env.TRAILING_SLASH === 'true'),
},
```

```diff
- <nuxt-link to="/imprint/">
+ <nuxt-link :to="'imprint' | urlHelper">
    Imprint
 </nuxt-link>

 <nuxt-link :to="blok.full_slug | urlHelper">
   {{ blok.name }}
 </nuxt-link>
```

Is this a viable solution? Maybe someone has another and better idea.

**Comment by larzon83** [comment]
> Is this a viable solution? Maybe someone has another and better idea.

I use the same approach.
One thing I did slightly different is the setting of`trailingSlash`. I have it like so:
```js
router: {
  trailingSlash:
    process.env.NUXT_ENV_IS_SPA === 'true' ||
    process.env.NODE_ENV === 'development'
      ? undefined
      : true
}
```

I set `trailingSlash` to `undefined` when I'm in development or when `NUXT_ENV_IS_SPA` is true (which is for my second version of the app for Storyblok).
`undefined` is the default in nuxt for `trailingSlash` -> routes work with and without trailing slashes.
In your code, for the storyblok version, `trailingSlash ` would result in `false`, which makes routes only work without trailing slashes.
With `undefined` I can leave my links with trailing slashes (`<nuxt-link to="/imprint/">`) and it still works. Also when opening a url inside storyblok without the trailing slash...

**Comment by marvinhuebner** [comment]
Thanks for your feedback. Then I'll just assume that this is somehow a common method to do.

> With `undefined` I can leave my links with trailing slashes (`<nuxt-link to="/imprint/">`) and it still works. Also when opening a url inside storyblok without the trailing slash...

I must check if this. For me it always threw errors because of the static mode, but I never set the `trailingSlash` option to `undefined` but to `false`. Thanks for the tip!

**Comment by studiolathe** [comment]
Hi @marvinhuebner @larzon83 

Sorry to rehash an old issue but I seem to be facing the same problem for a current build and was hoping you might be able to shed some light. I too am deploying to Netlify with Nuxt trailing slash activated and am also taking advantage of the Netlify's pretty URL function.

I have implemented your approach by setting a new `TRAILING_SLASH` env variable and contextually setting the `trailingSlash` but the Live Preview editor still won't work when using statically generated sites from Netlify but works fine locally. I can select content block in the preview but when I type in the SB content field there are no live changes? 

For context I have tried creating a new Netlify instance without the pretty URL settings selected and another approach using the [new branch specific vars in Netlify](https://docs.netlify.com/environment-variables/overview/) with the hope I can setup the `staging` branch previews to _not_ use trailing slashes and the production branch _to_ use them. With this approach we I have also added the below settings in my `netlify.toml` to ignore the pretty URLs post-processing feature from the staging branch build:
```
[context.staging.processing]
  skip_processing = true
```

But unfortunately neither seem to work. I am wondering either you might have some ideas as to what I am missing?

Thanks in advance, 
Sam

**Comment by larzon83** [comment]
I suspect static generation is the wrong target here. Try with ssr: false

**Comment by studiolathe** [comment]
@larzon83 that seemed to do the trick, thank you!

This leads me to wonder what is the best approach to workflows i.e. Staging > Production? Is there a way to use a webhook to rebuild the staging site on Netlify and then use another webhook to rebuild the production environment only when Published?

**Comment by marvinhuebner** [comment]
@studiolathe you should just deliver your staging environment as a single page app, then you don't need a rebuild when content is saved. The storyblok webhook then triggers your production build when something is published.

**Comment by studiolathe** [comment]
Thanks, @marvinhuebner 

Is there any chance would you be able share an example `nuxt.config.js` and a sample page request?

When setting `ssr.false` in dev the live editor is now working on my staging build but when creating a new page and making a request to it I'm getting `cv=undefined` added to the network request and `['This record could not be found']` in console?
 
```
// nuxt.config.js
...
ssr: process.env.TRAILING_SLASH === 'false' || process.env.NODE_ENV === 'development' ? false : true,

buildModules: [
  ["@storyblok/nuxt-2/module", {
    accessToken: process.env.STORYBLOK_STAGING_API_TOKEN, 
    bridge: true,
    useApiClient: true,
    apiOptions: {
      cache: {
        type: process.env.TRAILING_SLASH === 'false' || process.env.NODE_ENV === 'development' ? 'memory' : 'none',
      }
    },
  }],
  '@nuxtjs/composition-api/module',
],


// _slug.vue
...
async asyncData(context) {
    const storyblokApi = useStoryblokApi();
    let story = null;
    let relatedArticles = null;
    story = await storyblokApi
      .get(`cdn/stories/news/${context.params.slug}`, {
        version: "draft",
      })
      .then((res) => {
        return res.data.story;
      })
      .catch((res) => {
        if (!res.response) {
          console.error(res);
          context.error({
            statusCode: 404,
            message: "Failed to receive content form api",
          });
        } else {
          console.error(res.response.data);
          context.error({
            statusCode: res.response.status,
            message: res.response.data,
          });
        }
      });
    return {
      story
    };
...
```

Appreciate any help in advance!

**Comment by larzon83** [comment]
Creating a new page in Storyblok? Try reload the page.

**Comment by studiolathe** [comment]
@larzon83 yes, creating a new child article page or loading any child article inside Storyblok folder (routes to `pages/news/_slug.vue` in Nuxt) using the code the above seems to throw the error.

To be clear I'm using a Preview token and requesting `version.draft`, so I'm not sure why it would be throwing the error? Any thoughts?

**Comment by larzon83** [comment]
No clue sorry. Long time not worked on my project. Also Long time not updated deps. I probably use an older version of the storyblok integration and no nuxt bridge. Maybe ask in storyblok Discord?
My project works similar for dynamic pages, reading in context.params.slug and it works...

**Comment by lisilinhart** [comment]
Hi @marvinhuebner, is this still an issue?

**Comment by marvinhuebner** [comment]
Hi @lisilinhart, I'm still using my workaround described here: https://github.com/storyblok/storyblok/issues/613#issuecomment-788009337

I have not changed it since then.

**Comment by js-ms** [comment]
@jessicaribeiroalves have you found any solution? im struggling with the same thing

**Comment by jessicaribeiroalves** [comment]
> @jessicaribeiroalves have you found any solution? im struggling with the same thing

I haven't to be honest. My workaround was clicking to edit the story => click on **"Entry configuration"** (the middle icon on the top of the page) =>change the **"Real path"** to be the url with trailing slash
After doing that, the preview mode works again, but it's annoying because I need to change the real path for all my stories. Definitely not a solution.

**Comment by alvarosabu** [comment]
Hi @marvinhuebner, we are closing this issue because this repository was used in the early days for support-related questions, but we've since transitioned to more effective channels for helping our users.

Please, refer to the following resources:

Storyblok Github Org: https://github.com/storyblok
Storyblok Docs: https://www.storyblok.com/docs
Official Support: https://support.storyblok.com
Join our Discord: https://www.storyblok.com/join-discord

Thanks for the understanding 🙏🏻
